File: STRINGS/8. String to Integer (atoi).cpp
//problem link: https://leetcode.com/problems/string-to-integer-atoi/
//timeComplexity: O(n)
//spaceComplexity: O(1)

class Solution {
public:
    int myAtoi(string s) {
        bool sign=true;
        //removeSpace
        int indx=0;
        int n=s.size();
        while(indx<n && s[indx]==' '){
            indx++;
        }
        //signedness
        if(s[indx]=='-'||s[indx]=='+'){
            if(s[indx]=='-'){
                sign=false;
            }
            indx++;
        }
        //conversion
        long long sum=0;
        while(indx<n && isdigit(s[indx])){
            int x=s[indx]-'0';
            sum=sum*10+x%10;
            if(sum>INT_MAX){
                return sign==true?INT_MAX:INT_MIN;
            }
            if(sum<INT_MIN){
                return sign==true?INT_MAX:INT_MIN;
            }
            indx++;
        }
        return sign==true?sum:-sum;
    }
};

File: STRINGS/8. String to Integer (atoi).java
//problem link: https://leetcode.com/problems/string-to-integer-atoi/description/
//timeComplexity: O(n)
//spaceComplexity: O(1)

class Solution {
    public int myAtoi(String s) {
        int n=s.length();
        if(n==0)return 0;
        int indx=0;
        while(indx<n && s.charAt(indx)==' '){
            indx++;
        }
        boolean sign=true;
        if(indx<n && (s.charAt(indx)=='+'||s.charAt(indx)=='-')){
            sign=s.charAt(indx)=='+'?true:false;
            indx++;
        }
        long sum=0;
        while(indx<n && Character.isDigit(s.charAt(indx))){
            char ch=s.charAt(indx);
            int val=(int)(ch-'0');
            sum=sum*10+val;
            if(sum>Integer.MAX_VALUE){
                return sign==true?Integer.MAX_VALUE:Integer.MIN_VALUE;
            }
            indx++;
        }
        return sign==true?(int)sum:(int)(0-sum);
    }
}